### ISSUE TITLE
Select in certain cases is run after enumerator disposal

### ISSUE THREAD

**Issue by BratWiekszy** [dotnet/corefx#39192]
Consider following invocations on the same enumerable (enumerator):  
```c#
1: enumerable.Where((_, i) => i == 4).Select(x => x.Foo()).First();
2: enumerable.Select(x => x.Foo()).ElementAt(4);
3: enumerable.Skip(4).Select(x => x.Foo()).First();
4: enumerable.Select(x => x.Foo()).Take(5).Last();
5: enumerable.Take(5).Select(x => x.Foo()).Last();
```

All of them should return the same element with Foo() selector applied, and it happens. Also some of them use optimisations, which limit number of calls of selector (to 1, by calling only on 5th):
  
 - 1: limit exists and it's desired,
 - 2: limit does not exist, and it's weird because 5. has limit, even tho 2. is more suitable for such optimisation, but it's not a problem,
 - 3: limit exists and it's desired,
 - 4: limit does not exist but it's fine,
 - 5: limit exists but it's fine.

 In case of 3. and 5. selector function is run after a call to `Dispose()` on enumerator and it's problematic because it does not comply will exprected behavior of something disposable - it usually is usuable and its data unavailable. Also, in my case, enumerator fetches data headers in MoreNext() before returning, and in next call it decides whether to fetch full data based on call in Foo(). When no next call is made, this fetching is done in Dispose, at last point where it could be done, and then data source is closed. Even if i put full fetching code inside Foo(), it being called after Disposal completely invalidates this workaround.
 
Since now I know about this quirk, i can fix my calls, but consumers of my code may not have even slightest idea why their data is missing. I see two possible fixes:
  
 - Althought implementation of SelectIPartitionIterator, EnumerablePartition and respective TryGetLast() or TryGetFirst() methods of Enumerable are negligible detail, it does influence behavior of calls, and as such it should be mentioned somewhere in docs.
 - TryGetLast<T>(IEnumerable<T>, out bool) and TryGetFirst<T>(IEnumerable<T>, out bool) code that now looks like:
 ```c#
if (source is IPartition<TSource> partition)
{
    return partition.TryGetLast(out found);
}
``` 
could be replaced with something like this:
```c#
if (source is IPartition<TSource> partition)
{
    var result = partition.TryGetLast(out found);
    Dispose();
    return result;
}
```
and IPartition<T> methods like TryGetLast() and TryGetFirst() shouldn't dispose own enumerator before their dispose is called.

**Comment by stephentoub** [comment]
Can you please share a standalone repro of the specific Dispose problem you're concerned about? It's not clear to me from your example. Thanks.

**Comment by BratWiekszy** [comment]
[NeglectedDisposal.zip](https://github.com/dotnet/corefx/files/3360260/NeglectedDisposal.zip)

**Comment by Clockwork-Muse** [comment]
Tl;dr: The code you've provided is doing some rather unusual things, and should probably be changed, unless you can show a much more compelling case for _why_ the code should be written this way.

<hr />

You have a bunch of things going on here:
``` csharp
public class Enumerator : IEnumerable<DataObject>, IEnumerator<DataObject>
{
	private int _state;

	/*
	 * lots of stuff removed
	 */ 

	public IEnumerator<DataObject> GetEnumerator()
	{
		if(_state != -1)
			throw new InvalidOperationException();

		_state = 0;
		return this;
	}

	IEnumerator IEnumerable.GetEnumerator()
	{
		return GetEnumerator();
	}
}
```
... the enumerable can't be iterated over more than once, which is going to surprise essentially everybody.  Unless you want to spend the effort the duplicate the internal-only [Iterator type](https://referencesource.microsoft.com/#System.Core/System/Linq/Enumerable.cs,00f3550fbfefd123), I recommend you split them up, and make sure to return a new `Enumerator` each call.

<hr />

```csharp
// Some of your test code
[Test]
public void Take_Select_Last()
{
	var ds = new DataSource();
	var enumerable = new Enumerator(ds);
	var dataObject = enumerable.Take(3).Select(o => o.Fetch("some-prop")).Last();

	Assert.True(ds.Log.Contains("_fetch_dispose"));
}

// Inside your Enumerator
public void Dispose()
{
	if(_current?.Included == true)
		_source.FetchProperties(_current);

	_source.Dispose();
}
```
...  You're dealing with what's known as a "lifetime" problem.  Specifically, your `DataObject`s are attempting to outlive your `DataSource`, because your enumerator is disposing your source.  What if I want to find out multiple things?  Eg:

```csharp
public void SomethingElse()
{
	var ds = new DataSource();
	var enumerable_1 = new Enumerator(ds);
	var dataObject = enumerable_1.Take(3).Select(o => o.Fetch("some-prop")).Last();

	var enumerable_2 = new Enumerator(ds);
	// Oops - ds has been disposed!
	var dataObject2 = enumerable_2.Skip(3).Select(o => o.Fetch("some-prop")).First();
}
```
... You probably shouldn't be disposing the datasource inside the enumerator.  Also, calling non-cleanup methods in a `Dispose` is a pretty big no-no, especially something that's **getting additional information to act on**.

<hr />

You have a strange two-stage portion to your enumerator:
```csharp
// Enumerator starts with _state == 0
public bool MoveNext()
{
	switch (_state)
	{
		case 0:
			if ((_current = _source.FetchNext()) != null)
			{
				_state = 1;
				return true;
			}
			break;
		case 1:
			if (_current?.Included == true)
			{
				_source.FetchProperties(_current);
			}
			goto case 0;
	}
	return false;
}
```

... `_source.FetchProperties(_current);` isn't getting called until the **NEXT** time `MoveNext()` is called (and, as you've noticed, only if `dataObject.Fetch()` was being called).  You never have a guarantee that `MoveNext()` will ever be called after the "current" element - `First()`, especially, is going to ignore subsequent elements, and `Take()` just keeps an integer count.  This isn't even going into more involved optimizations involving arrays or partitions.  If you need additional data, you're going to have to get it either during `MoveNext()`, or make sure that your datasource lives long enough, and fetch it directly during a subsequent operation (say, directly on the object result of `First()` or `Last()`):

```csharp
public void Example()
{
	using (var ds = new DataSource())
	{
		var enumerable = new Enumerator(ds);
		var dataObject = enumerable.Take(3).Last();
		ds.FetchProperties(dataObject);
		// do something with dataObject here.
		// Returning dataObject from this method is problematic if it keeps a reference to ds
	}
}
```
... it would also be returning that data way too late to use in other types of queries:

```csharp
public void Example()
{
	using (var ds = new DataSource())
	{
		var enumerable = new Enumerator(ds);
		var dataObjects = enumerable.Where(o => 
				{
					o.Fetch("some-prop");
					// But some-prop won't be retrieved until MoveNext() is called?
					return o.Get("some-prop") == "whatever";
				}).Count();
	}
}
```

If you're trying to do something like batch retrieval of expensive properties, you might be better off having a decorator for the `DataSource` that spins up an enumerator that gets just those properties during `MoveNext()`.

**Comment by BratWiekszy** [comment]
Well, I'm kinda surprised of such elaborative answer, explaining why _example_ repro code is wrong. This strange two-stage portion was initially a iterator method, with even stranger code inside. I'm not changing it for reasons, but in this case it isn't really a problem. The data source is stream-ish, and client are perfectly aware that it can't be enumerated twice. That's why there's whole linq pipeline to select whatever they want. Also I don't really care for what happens to dataObject equivalents, because they are readonly and are their purpose is to just carry the damn data. DataSource object really is internal wrapper for reader so there's a point in disposing it right away. 
Whole api is generally used in slightly different manner, so I only learned about this method-combination problem by accident, and it's not like _without_ the fix some kind of tragedy happens, it doesn't. I already got a workaround for that. But i like the fact that there's a concern to how this code is written, but you not seem to bat an eye for the inconsistency that this issue was meant to point out. I mean, that Dispose method is there for a reason. If it shouldn't be used, why it's there in first place? And, if it's already there and it won't be moving away soon, it seems perfectly fine to utilize it. If you're all certain that this shouldn't be discussed, just close the issue, that woulda save y'all lots of otherwise pointless typing.

**Comment by mburbea** [comment]
I don't think its an unusual for an enumerator to not be usable after being enumerated once. Resharper warns you about this behavior. But here's a minimal repro, something that reads from a stream and the `enumerable` is disposed after reading.

I tried all 5 scenarios and it didn't seem to throw an exception for me (.net core 2.2)

```cs
private static void TestEnumerables()
{
    IEnumerable<byte> OneTimeIterable()
    {
        var m = new MemoryStream(new byte[] { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 });
        return Inner();
        IEnumerable<int> Inner()
        {
            using (m)
            {
                for (int i = 0; i < 10; i++)
                {
                    yield return m.ReadByte();
                }
            }
        }
    }

    OneTimeIterable().Where((_, i) => i == 4).Select(x => x.ToString()).First();
    OneTimeIterable().Select(x => x.ToString()).ElementAt(4);
    OneTimeIterable().Skip(4).Select(x => x.ToString()).First();
    OneTimeIterable().Select(x => x.ToString()).Take(5).Last();
    OneTimeIterable().Take(5).Select(x => x.ToString()).Last();
}
```
My `OneTimeIterable()` will throw an exception if it's reused.
e.g.
```
var z = OneTimeIterable();
z.First();
z.First(); // throws a Stream is disposed exception
```

**Comment by Clockwork-Muse** [comment]
@mburbea 
> I don't think its an unusual for an enumerator to not be usable after being enumerated once.

`IEnumerator`, no.  `IEnumerable`, yes.  `IEnumerable`, itself, does not extend `IDisposable`.  You can have something that implements both (as, essentially, you do here), but many callers are going to look at you funny if you auto-drop the enumerable after enumeration finishes, even if enumeration will be expensive.  Your sample code is throwing an error only because the underlying stream was disposed (you changed the state of your enumerable), not because the enumerator was disposed.

**Comment by mburbea** [comment]
I disagree. Like I said, when I use [Resharper](https://www.jetbrains.com/help/resharper/PossibleMultipleEnumeration.html) it warns me when I try to reuse an `IEnumerable<T>` after it has already been enumerated.

For an example of something you'd see in the wild, Dapper's queries with the flag `buffer:false`, will be in the same peculiar affair of being able to read the content through once, but if you try to do it again it will throw an exception.

**Comment by Clockwork-Muse** [comment]
@BratWiekszy 

> But i like the fact that there's a concern to how this code is written, but you not seem to bat an eye for the inconsistency that this issue was meant to point out. I mean, that Dispose method is there for a reason. If it shouldn't be used, why it's there in first place? And, if it's already there and it won't be moving away soon, it seems perfectly fine to utilize it.

...because `Dispose()` _is not intended to be used this way_.  `Dispose()` is intended for release of any native resources your code might hold - period.  Code that performs useful work inside a `Dispose()` method is very surprising, and the normal assumption is going to be that it's misbehaving.

This code:

```csharp
if (source is IPartition<TSource> partition)
{
    var result = partition.TryGetLast(out found);
    Dispose();
    return result;
}
```
... literally cannot be written - there's not necessarily a `Dispose()` method on random `IEnumerable`s passed to `Last()` (and `IPartition` doesn't extend `IDisposable` either)

**Comment by Clockwork-Muse** [comment]
@mburbea -
>I disagree. Like I said, when I use Resharper it warns me when I try to reuse an IEnumerable<T> after it has already been enumerated.
> 
> For an example of something you'd see in the wild, Dapper's queries with the flag buffer:false, will be in the same peculiar affair of being able to read the content through once, but if you try to do it again it will throw an exception.

Of course Resharper warns you - there _are_ potential issues with iterating twice.  And almost all code that directly consumes an `IEnumerable` (as opposed to being part of the stream) saves the contents off.  However, that doesn't render the `IEnumerable` invalid, just memoizes the contents.

(I'd have to look more at Dapper, but at minimum it's a very constrained specific case, and a non-default option to boot)

**Comment by stephentoub** [comment]
FWIW, I found @BratWiekszy's repro helpful, and I agree it's a little unintuitive that we're invoking callbacks from within the query after the enumerator is disposed.  However, I also agree it's a very unusual use case, where a predicate depends on state that's mutated by the enumerator's dispose, and I believe we should reserve the flexibility to order execution in this manner, just as we reserve the right to skip invocation of some callbacks altogether.
